util/tracer_nvbit/tracer_tool/traces-processing/post-traces-processing.cpp

```diff
@@ -109,46 +113,76 @@ int main(int argc, char **argv) {
   ifstream ifs;
   ofstream ofs;
 
-  ifs.open(kernellist_filepath.c_str());
-  ofs.open((string(kernellist_filepath) + ".g").c_str());
-
-  if (!ifs.is_open()) {
-    cerr << "Unable to open file: " << kernellist_filepath << endl;
+  // We can now pass a directory or a file as the input argument 
+  std::filesystem::path p(filepath);
+  if (std::filesystem::is_directory(p))
+  {
+    for (const auto & entry : std::filesystem::directory_iterator(p))
+    {
+      std::string filename = entry.path().filename();
+      if (filename.find("kernelslist") != std::string::npos)
+      {
+        kernelslist_list.push_back(entry.path().string());
+      }
+    }
+  }
+  else if (std::filesystem::is_regular_file(p))
+  {
+    kernelslist_list.push_back(kernellist_filepath);
+  }
+  else
+  {
+    cerr << "Invalid file path\n";
     return 1;
   }
 
-  string directory(kernellist_filepath);
-  const size_t last_slash_idx = directory.rfind('/');
-  if (std::string::npos != last_slash_idx) {
-    directory = directory.substr(0, last_slash_idx);
-  }
+  for (auto kernellist_filepath : kernelslist_list){
+    string directory(kernellist_filepath);
+    const size_t last_slash_idx = directory.rfind('/');
+    if (std::string::npos != last_slash_idx) {
+      directory = directory.substr(0, last_slash_idx);
+    }
 
-  string line;
-  string filepath;
-  while (!ifs.eof()) {
-    getline(ifs, line);
-    if (line.empty())
-      continue;
-    else if (line.substr(0, 6) == "Memcpy") {
-      ofs << line << endl;
-    } else if (line.substr(0, 6) == "kernel") {
-      filepath = directory + "/" + line;
-      group_per_block(filepath.c_str());
-
-      int _l = line.length();
-      if (_l > 3 && line.substr(_l - 3, 3) == ".xz") {
-        ofs << line.substr(0, _l - 3) << "g.xz" << endl;
+    ifs.open(kernellist_filepath.c_str());
+    // If we have only one context, name it kernelslist.g by default
+    if(kernelslist_list.size() == 1 || kernelslist_list[0] == kernellist_filepath)

```

Comment:
Cool. not a lot of code to maintain backward compatibility.
However, we will have to update the simulation runner to handle workloads that actually have multiple contexts - but we can do that in another commit.

Reply:
Agree. For now, I think the ability to even trace them sanely is a good change.